### ISSUE TITLE
Proposal: Fluent class definition

### ISSUE THREAD

**Issue by alanhauge** [dotnet/csharplang#2976]
# Fluent Class Definition

* [x] Proposed
* [x] Prototype: [Proof-of-concept, beta version](https://github.com/GalvanizedSoftware/Beethoven)
* [ ] Implementation: _TODO_
* [ ] Specification: _TODO_

## Summary
[summary]: #summary

Fluent style programming applied to class definition.
Similar code is imported into the class multiple times rather than being coded multiple times.

## Motivation
[motivation]: #motivation

Writing code using composition over inheritance will give better code.
Having the possibility to import code into a class without mapping code, will give shorter and more readable code. 

When writing the same code over and over, there is a risk of copy-paste errors.
These errors a difficult to spot, and are normally caught by a person, who has not written the code.

Even though the code is repeatet, and is almost identical,
some people could add there own code to fix a specific bug.
The 'standard' implementation is proven to work, but adding code to it makes the results unpredictable.

In the longer run a runtime version will give a lot of possibilities.
The code can be compiled to fit runtime conditions not know at compile time.

If mapping or implementation is automatic, 
the class does not need to change just because the interface or lower level technology changes.

## Detailed design
[design]: #detailed-design

To introduce fluent code definition, an option to to extent the auto-property system.

A class might be defined as:
```CS
class ClassA : INotifyPropertyChanged
{
  ILogger _logger = Logger.FindLogger(nameof(ClassA));

  int Option
  {
    get;
    set
      .SkipIfSame()
      .Set()
      .NotifyPropertyChanged();
  }

  string GetDiscription(string header)
    .Log(_logger, nameof(GetDiscription), header)
    .SkipIf<string>( () => Option == 0, "Nothing")
    .Return<string>($"{header] is {Option}")

  event PropertyChangedEventHandler PropertyChanged;
}
```

This should be compiled into a full implementation, so the methods (SkipIfSame, Log, etc.) are only used at compile time.

To control the flow inside the property implementation, the imported code is using this signature:
```
bool Foo(ref int returnValue)
```

If this returns ```true```, the execution continues, if ```false``` the execution stops.
At the end, ```returnValue``` is returned.

For the method, the internal signature is:
```CS
bool GetDiscription(string header, ref string returnValue)
```

So the internal code, could look like this:
```CS
class ClassA : INotifyPropertyChanged
{
  ILogger _logger = Logger.FindLogger("ClassA");
  int _option;

  int Option
  {
    get { return _option; }
    set
    {
      if (value.Equals(_option))
        return;
   
      _option = value;

      PropertyChanged?.Invoke(this, new PropertyChangedEventArgs("Option"));
    }
  }

  string GetDiscription(string header)
  {
    string returnValue = default(string);

    _logger.WriteLog("GetDiscription", header);

    if (Option == 0)
    {
      returnValue = "Nothing";
      return returnValue;
    }

    returnValue = $"{header} is {Option}";
    if (true)
      return returnValue;

    return returnValue;
  }

  event PropertyChangedEventHandler PropertyChanged;
}
```
Alternatively, it might be nessesay to call methods instead of including the code.

### Types

Types for holding the defition are needed:
``` CS
TypeDefinition
GetPropertyDefinition
SetPropertyDefinition
MethodDefinition
EventDefinition
```

The definition types should be used internally in the compiler, but should also be public.
This will enable programmers to define classes runtime.

The dot-notation used in the class definition are extensions to a definition,
this also enables users to add their own. 

For example:
``` CS
SetPropertyDefinition SkipIfEqual(this SetPropertyDefinition previous)
{
  return new SetPropertyDefinition(previous, 
    (oldValue, newValue) => newValue.Equals(newValue))
}
```
(_Simplyfied for readability, this actual code might not work._)

### Default implementation
With the new notation, properties end up having an almost identical code.
So introducing a default implementation would result in less code.

The idea is to only define an interface, and the class just refers to a default implementation. For example:
``` CS
class ClassB : ISomeInterface
{
  import: MyDefaultPropertyImplementation defaultImplementation = new MyDefaultPropertyImplementation();
}
```

This means: Implement all properties in the interface using MyDefaultPropertyImplementation as a template.

I'm not sure this notation is the best way to do this,
but I don't want to mix it up with (_evil_) inheritance by having it one the class definition line,
and the compiler need to know somebody has implemented the interface, so it has to be a new keyword.

For cases not covered by the default implementation, custom implementation can be done inside the class.
``` CS
class ClassB : ISomeInterface
{
  import: MyDefaultPropertyImplementation defaultImplementation = new MyDefaultPropertyImplementation();

  void Foo2()
  {
    // Default implementation ignored, this is used instead
  }
}
```

The import feature could also be used to import partial implementation from another class.

``` CS
class ClassB : ISomeInterface
{
  import ISomeInterface: Foo1Implementation foo1 = new Foo1Implementation(); // Auto-map matching methods from ISomeInterface to class
}

class Foo1Implementation
{
  string Foo1()
  {
    return "1";
  }
}
```

### New posibilities

This tooling will open new possibilies:
* Much easier runtime compilation than existing emit-namespace
* Easier composition of classes where implementation is auto-mapped to good and **SOLID** implementations.
* Auto-mapping enables a form of duck-typing, auto-generated wrapper class.

## Drawbacks
[drawbacks]: #drawbacks

Major change, will have consequences in many places.

A fundametally different way of coding classes may confuse new programmers.

In some cases it might be unclear or illogical what version of the implementation is used.
This should be solve either by compiler errors or very clear rules.

This way of defining classes and code woold be confusing to some,
so the notation should be as self-explanatory as possible.

## Alternatives
[alternatives]: #alternatives

All features can be done with runtime compilation using the Roslyn runtim compilier.
However, performance would be better, if the ode was generated at compile time.

## Unresolved questions
[unresolved]: #unresolved-questions

Notation is a big issue, it will be a big task to find a way to introduce this without adding inconsistency to the language.

**Comment by YairHalberstadt** [comment]
I think the desire for easy ways to compose bits of code macro style makes a lot of sense.

However this proposal is very difficult to assess, since it doesn't formally describe how a eg. a `SetPropertyDefinition` is defined.

I think what would be really useful is providing a formal description of `SetPropertyDefinition`, and then defining all three of:

SkipIfSame
Set
NotifyPropertyChanged

I'm interested in how that's done.

**Comment by alanhauge** [comment]
A (simplified) version of SetPropertyDefinition would look like this:
``` csharp
  public abstract class SetPropertyDefinition<T> : SetPropertyDefinition
  {
    private readonly ISetPropertyDefinition<T>[] definitions;

    protected SetPropertyDefinition(string name)
    {
      Name = name;
      definitions = new ISetPropertyDefinition<T>[0];
    }

    protected SetPropertyDefinition(PropertyDefinition previous, ISetPropertyDefinition<T>[] propertyDefinitions) :
      this(previous?.Name)
    {
      definitions = previous?
        .definitions
        .Concat(propertyDefinitions)
        .ToArray();
    }

    public string Name { get; }

    public bool Invoke(T newValue)
    {
      foreach (IPropertyDefinition<T> definition in definitions)
        if (!definition.InvokeSetter(newValue))
          return false;
      return true;
    }
}
```
The ```Invoke``` method would only be used in a runtime scenario, I imagine this code would be generated be the compiler in normal cases.

Anyone implementing a code __macro__ would have to implement:
``` csharp
  public interface ISetPropertyDefinition<T>
  {
    bool Invoke(T newValue);
  }
```

For example:
```csharp
  internal class SkipIfEqual<T> : ISetPropertyDefinition<T>
  {
    private T oldValue;

    public bool Invoke(IT newValue)
    {
      if (newValue == null)
      {
        if (oldValue == null)
          return false;
      }
      else if (newValue.Equals(oldValue))
        return false;
      oldValue = newValue;
      return true;
    }
  }
```
And the code fluent-style addition ```.SkipIfEqual()``` is done with an extension:
```csharp
    public static SetPropertyDefinition<T> SkipIfEqual<T>(this SetPropertyDefinition<T> propertyDefinition) => 
      new SetPropertyDefinition<T>(propertyDefinition, new SkipIfEqual<T>());
```

```Set``` would be done by linking setter to the getter, making sure the set value returned by the getter. I cannot easily paste in an example because in my implementation, both property get and set are handled as one type. So implementing a ```GetterSetter``` is very simple.

```NotifyPropertyChanged``` is a little complex to paste in here.
I introduced binding between the main instance and the implementor (```ISetPropertyDefinition```), so the event was triggered by the implementor and it passes the main instance in ´´´PropertyChanged´´´. 
I think part of the complexity comes in because I implemented a runtime version.
I also needed to implement my own handling of events, but compile time, it would be much simpler to do all this.

I did a proof-of-concept of a runtime version, called [Beethoven](https://github.com/GalvanizedSoftware/Beethoven), you (or anyone else readingthis) is interested.

**Comment by HaloFour** [comment]
This seems really confusing, particularly where the compiler needs to switch between compile-time macros and runtime execution.  It involves basically creating an entire new language within a language.

All of the use cases described fit source generators perfectly, using existing syntax and attributes instead of DSLs to describe the behavior and allowing for analyzers to inject the necessary code to intercept and inject the configured behavior.

**Comment by alanhauge** [comment]
Sorry for the confusion.
I don't believe the switch between compiler generated code and runtime generated code is more complicated than what is already in auto-properties.

The reason this is suggested as a feature in the language, not a purely external component is that there are so many ways and implementations of doing this (or parts of it) already.
I think it's time to align all this to a solution within the language, that is *not* domain specific..

**Comment by HaloFour** [comment]
> there are so many ways and implementations of doing this (or parts of it)

IMO that's a good reason to keep it out of the language and support it through analyzers.  The many different implementations have their own policies and configuration which is best left to libraries which enable quick evolution and customization.  If implemented within the language the team would choose their idea of the idiomatic approach and that would be it.

**Comment by mv10-work** [comment]
This almost looks like LINQ for property backing-fields. Seems like it would be a lot easier to just give us access to the backing fields so we can write code like "skip if same" logic ourselves when needed (or any of the other backing-field requests that have piled up over the years, like being able to declare it explicitly).

**Comment by HaloFour** [comment]
> like being able to declare it explicitly

Wouldn't that then be a regular non-auto property?

**Comment by mv10-work** [comment]
> Wouldn't that then be a regular non-auto property?

Generally I just meant there have been a lot of requests related to accessing the backing fields. I can't find it now but I seem to remember somebody asking for something along these lines:

```csharp
public var foo bool YesOrNo { get { ... } set { ... } }
```

where `foo` becomes the backing-field within `get` and `set`. I probably have the proposed syntax wrong though, it seems like it was a couple years ago.

**Comment by quinmars** [comment]
This is probably the moment to mention #140. :smirk:

**Comment by HaloFour** [comment]
And probably #133.

**Comment by alanhauge** [comment]
> This almost looks like LINQ for property backing-fields. Seems like it would be a lot easier to just give us access to the backing fields so we can write code like "skip if same" logic ourselves when needed (or any of the other backing-field requests that have piled up over the years, like being able to declare it explicitly).

Giving access to the backing field would not "solve" the issue  or make this feature irrelevant.
The point is to be able to import good code, that is proven to work, so we are not writing repetitive code with the risk of copy-paste errors.

The reason I bring up auto-properties is that it's something we trust to do its job correctly, always.
We would never consider that the auto-property is not working, not because we are ignorant, but because it is proven to work.

I like to write the "skip if same" code once and then import it everywhere I need it, without the restrictions of inheritance, an additional field, injected code from 3rd party component etc.
When I see SkipIfSame 50 times, I known the code is doing the same thing. If I see 50 similar if-returns, I'm not 100% sure they are all doing the same thing simply because there is more code to compare.

**Comment by sharwell** [comment]
@alanhauge The proposals like `field` would allow something roughly like this:

```
int Option {
  get;
  set => SetAndNotifyIfChanged(ref field, value);
}
```

The `SetAndNotifyIfChanged` method can be defined in an abstract base class for supporting implementations of `INotifyPropertyChanged`.

**Comment by alanhauge** [comment]
My suggestion is not about solving the specific problem with setting changed properties, accessing the inner backing fields or introducing a DSL for that particular purpose.

It's about importing external code into a class in a simpler and more decoupled way than the current solutions.

I made created another example of how the code behind this class would work:
```csharp
class ClassA : INotifyPropertyChanged
{
  ILogger _logger = Logger.FindLogger(nameof(ClassA));

  int Option
  {
    get;
    set
      .SkipIfSame()
      .Set()
      .NotifyPropertyChanged();
  }

  string GetDiscription(string header)
    .Log(_logger, nameof(GetDiscription), header)
    .SkipIf<string>( () => Option == 0, "Nothing")
    .Return<string>($"{header] is {Option}")

  event PropertyChangedEventHandler PropertyChanged;
}
```

The most relevant file is the code behind ClassA:
```csharp
  public class ClassA : INotifyPropertyChanged
  {
    private readonly ILogger _logger = Logger.FindLogger(nameof(ClassA));
    int backing_field_Option;
    ISetPropertyDefinition<int> import_backing_set_Option1;
    ISetPropertyDefinition<int> import_backing_set_Option2;
    ISetPropertyDefinition<int> import_backing_set_Option3;
    IMethodDefinition<string> import_backing_method_GetDiscription1;
    IMethodDefinition<string> import_backing_method_GetDiscription2;
    IMethodDefinition<string> import_backing_method_GetDiscription3;

    public ClassA()
    {
      import_backing_set_Option1 = new SkipIfEqual<int>();
      import_backing_set_Option2 = new SetProperty<int>(value => backing_field_Option = value);
      import_backing_set_Option3 = new NotifyPropertyChanged("Option", this, PropertyChanged);
      import_backing_method_GetDiscription1 = new LoggerDefinition(_logger, "GetDiscription");
      import_backing_method_GetDiscription2 = new SkipIfMethod<string>(() => Option == 0, "Nothing");
      import_backing_method_GetDiscription3 = new Return<string, string>(header => $"{header} is {Option}");
    }

    public int Option
    {
      get => backing_field_Option;
      set
      {
        if (!import_backing_set_Option1.Invoke(value))
          return;
        if (!import_backing_set_Option2.Invoke(value))
          return;
        if (!import_backing_set_Option3.Invoke(value))
          return;
      }
    }

    public string GetDiscription(string header)
    {
      string returnValue = null;
      object[] parameters = new object[] { header };
      if (!import_backing_method_GetDiscription1.Invoke(parameters, ref returnValue))
        return returnValue;
      if (!import_backing_method_GetDiscription2.Invoke(parameters, ref returnValue))
        return returnValue;
      if (!import_backing_method_GetDiscription3.Invoke(parameters, ref returnValue))
        return returnValue;
      return returnValue;
    }

    public event PropertyChangedEventHandler PropertyChanged;
  }
```
It's not the intention anyone would ever see this code, it's only included to illustrate what goes on behind the scenes.

All other files are here: https://github.com/GalvanizedSoftware/FluentClassDefinition

The bigger value of the feature is not in this example, it comes when whole classes are imported (auto-mapped) or a whole class is getting its implementation from an import statement.
In the last case, a change to the interface will not require any change to the class.

**Comment by HaloFour** [comment]
@alanhauge

> It's about importing external code into a class in a simpler and more decoupled way than the current solutions.

That's also the goal of source generators:  https://github.com/dotnet/csharplang/issues/107

The advantages of source generators is that they require very few changes to the language, and they open up lots of additional use cases such as writing type providers.

**Comment by alanhauge** [comment]
When you mentioned source generators first, I thought you meant the existing tools like for the entity framework or resource class generators. I did not realize it was a feature (almost) in the pipeline.

I see a lot of the same neat possibilities in there as I discovered when I did my proof-of-concept. Although I approched the problem from a different angle, my gutt-fealing is that the core feature is the same.

I will look into it some more.
